addons/portal/controllers/portal.py

```diff
@@ -196,6 +194,82 @@ def account(self, redirect=None, **post):
         response.headers['X-Frame-Options'] = 'DENY'
         return response
 
+    @http.route('/portal/attachment/add', type='json', auth='public')
+    def attachment_add(self, name, data, res_model, res_id, access_token=None, **kwargs):
+        """Process a file uploaded from the portal chatter and create the
+        corresponding `ir.attachment`.
+
+        The attachment will be created "pending" until the associated message
+        is actually created, and it will be garbage collected otherwise.
+
+        :param name: name of the file to save.
+        :type name: string
+
+        :param data: content of the file to save, base64 encoded.
+        :type data: string or bytes
+
+        :param res_model: name of the model of the original document.
+            To check access rights only, it will not be saved here.
+        :type res_model: string
+
+        :param res_id: id of the original document.
+            To check access rights only, it will not be saved here.
+        :type res_id: int
+
+        :param access_token: access_token of the original document.
+            To check access rights only, it will not be saved here.
+        :type access_token: string
+
+        :return: attachment data {id, name, mimetype, file_size, access_token}
+        :rtype: dict
+        """
+        try:
+            self._document_check_access(res_model, int(res_id), access_token=access_token)
+        except (AccessError, MissingError) as e:
```

Comment:
For the record `consteq` (from passlib) can also throw a `TypeError` when the operands do not have the same type. This can occur when the document does not have a token, and the crash itself gives some info to the caller. Perhaps the simplest would be to have `_document_check_access` check that the document has a token before comparing it?

Reply:
On the current code of master, `_document_check_access` already has that condition: `if not access_token or not consteq(document_sudo.access_token, access_token):`.

Reply:
> I suppose we can add `or not document_sudo.access_token` there.

yes, that's what I meant :)

Reply:
Done